src/main/java/com/android/volley/toolbox/AsyncHttpStack.java

```diff
@@ -0,0 +1,113 @@
+package com.android.volley.toolbox;
+
+import androidx.annotation.Nullable;
+import com.android.volley.AuthFailureError;
+import com.android.volley.Request;
+import com.android.volley.VolleyLog;
+import java.io.IOException;
+import java.util.Map;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.atomic.AtomicReference;
+
+/** Asynchronous extension of the {@link BaseHttpStack} class. */
+public abstract class AsyncHttpStack extends BaseHttpStack {
+
+    public interface OnRequestComplete {
+        /** Invoked when the stack successfully completes a request. */
+        void onSuccess(HttpResponse httpResponse);
+
+        /** Invoked when the stack throws an {@link AuthFailureError} during a request. */
+        void onAuthError(AuthFailureError authFailureError);
+
+        /** Invoked when the stack throws an {@link IOException} during a request. */
+        void onError(IOException ioException);
+    }
+
+    /**
+     * Makes an HTTP request with the given parameters, and calls the {@link OnRequestComplete}
+     * callback, with either the {@link HttpResponse} or error that was thrown.
+     *
+     * @param request to perform
+     * @param additionalHeaders to be sent together with {@link Request#getHeaders()}
+     * @param callback to be called after retrieving the {@link HttpResponse} or throwing an error.
+     */
+    public abstract void executeRequest(
+            Request<?> request, Map<String, String> additionalHeaders, OnRequestComplete callback);
+
+    /**
+     * Performs an HTTP request with the given parameters.
+     *
+     * @param request the request to perform
+     * @param additionalHeaders additional headers to be sent together with {@link
+     *     Request#getHeaders()}
+     * @return the {@link HttpResponse}, or null if an InterruptedException occurs.
```

Comment:
I don't think callers expect null responses, do they? We can't break the base interface we're implementing here by doing something undocumented/unexpected there. InterruptedIOException might be a reasonable choice here.

Reply:
Okay, that should be a better option.